File: 2nd-try/209/209.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <limits>

using namespace std;

class Solution
{
public:
    int minSubArrayLen(int target, vector<int> &nums)
    {
        // 思路：尝试快慢双指针，构成一个滑动窗口，动态更新最短长度
        int slow = 0, fast = 0;
        const int MAX = numeric_limits<int>::max();
        int minLen = MAX;
        int curSum = nums[fast];
        while (1)
        {
            if (curSum >= target)
            {
                minLen = min(minLen, fast - slow + 1);
                curSum -= nums[slow];
                slow++;
            }
            else
            {
                fast++;
                if (fast >= nums.size())
                    break;
                curSum += nums[fast];
            }
        }

        return minLen == MAX ? 0 : minLen;
    }
};

int main()
{
    vector<int> nums = {2,
                        3,
                        1,
                        2,
                        4,
                        3};
    Solution obj;
    int minLen = obj.minSubArrayLen(7, nums);
    cout << minLen << endl;

    return 0;
}

File: Array/209-the_shortest_sub-array/209.py
from typing import List

class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        start = 0
        end = 0
        sum_nums = nums[end]  # 先加一个，这样后面就可以避免“先累加再自增”的顺序
        count_ans = len(nums) + 1
        count_temp = 0
        while 1: # 后面有跳出判断了
            if sum_nums >= target:
                count_temp = end - start + 1 # 要+1
                if count_temp < count_ans:
                    count_ans = count_temp
                sum_nums -= nums[start]
                start += 1

            else:
                # 这里就可以先自增，再判断要不要跳出，再累加
                end += 1
                if end >= len(nums):
                    break
                sum_nums += nums[end]

        if count_ans > len(nums):
            return 0
        return count_ans
    
obj = Solution()
res = obj.minSubArrayLen(7, [2, 3,1,2,4,3])